File: src/operators/ReduceSum.cpp
// Copyright 2018 The AITS DNNC Authors.All Rights Reserved.
//
// Licensed to the Apache Software Foundation(ASF) under one
// or more contributor license agreements.See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.See the License for the
// specific language governing permissionsand limitations
// under the License.
//
// This file is part of AITS DNN compiler maintained at
// https://github.com/ai-techsystems/dnnCompiler
//

#include "operators/ReduceSum.h"

using namespace dnnc;
using namespace Eigen;

#ifdef DNNC_REDUCESUM_TEST
#include <iostream>
int main() {
  float f1[6] = {1., 2., 3., 4., 5., 6};
  float f2[9] = {6., 1., 0., 5., 7., 4., 2., 3., 8};

  tensor<float> tf1({2, 3});
  tf1.load(f1);
  tensor<float> tf2({9});
  tf2.load(f2);

  std::vector<int> axes({0});

  ReduceSum<int, float> sum("floatsum");

  sum.setAttribute(attr_axis, axes);
  sum.setAttribute(attr_keepdims, 0);

  std::cout << sum.compute(tf1) << std::endl;
  std::cout << sum.compute(tf2) << std::endl;
}
#endif

File: test/swig/ReduceSum.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
# pylint: disable=invalid-name, unused-argument
#
# This file is part of DNN compiler maintained at
# https://github.com/ai-techsystems/dnnCompiler

import common # DNNC path setup

import deepC.dnnc as dc
import numpy as np
import unittest



class ReduceSumTest(unittest.TestCase):
    def setUp(self):
        self.len = 24
        self.np_a = np.random.randn(self.len).astype(np.float32)
        self.dc_a = dc.array(list(self.np_a))

    def test_ReduceSum1D (self):
        npr = np.sum(self.np_a, axis=(0))
        dcr = dc.reduce_sum(self.dc_a, [0])
        np.testing.assert_allclose(npr, np.array(dcr.data()).astype(np.float32),
                rtol=1e-3, atol=1e-3)

    def test_ReduceSum2D (self):
        np_a = np.reshape(self.np_a, (6,4))
        dc_a = dc.reshape(self.dc_a, (6,4))

        npr = np.sum(np_a, axis=(0))
        dcr = dc.reduce_sum(dc_a, [0])
        np.testing.assert_allclose(npr.flatten(), np.array(dcr.data()).astype(np.float32),
                rtol=1e-3, atol=1e-3)

    def test_ReduceSum3D (self):
        np_a = np.reshape(self.np_a, (2,4,3))
        dc_a = dc.reshape(self.dc_a, (2,4,3))

        npr = np.sum(np_a, axis=(0))
        dcr = dc.reduce_sum(dc_a, [0])
        np.testing.assert_allclose(npr.flatten(), np.array(dcr.data()).astype(np.float32),
                rtol=1e-3, atol=1e-3)

    def test_ReduceSum4D (self):
        np_a = np.reshape(self.np_a, (2,2,3,2))
        dc_a = dc.reshape(self.dc_a, (2,2,3,2))

        npr = np.sum(np_a, axis=(0))
        dcr = dc.reduce_sum(dc_a, [0])
        np.testing.assert_allclose(npr.flatten(), np.array(dcr.data()).astype(np.float32),
                rtol=1e-3, atol=1e-3)

    def tearDown(self):
        return "test finished"

if __name__ == '__main__':
    unittest.main()

